File: backend/src/mapping_store.py
import json
import logging
import os

import psycopg
from psycopg.types.json import Jsonb

logger = logging.getLogger(__name__)

# Bundled preset shipped in the repo — seeded into the table once so a fresh
# database starts with the same default mapping the app has always shipped.
_DEFAULT_MAPPING_NAME = "electra_default"
_DEFAULT_MAPPING_PATH = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), "..", "mappings", f"{_DEFAULT_MAPPING_NAME}.json"
)

# Set once per process (e.g. once per warm serverless container) so the
# schema/seed check doesn't run on every single request.
_schema_ready = False


def _connect():
    """
    Opens a fresh connection per call — the right pattern for serverless,
    where the process itself is short-lived and DATABASE_URL is expected to
    point at a pooled (e.g. PgBouncer) connection string. Used as a context
    manager, the connection commits and closes automatically on exit.
    """
    database_url = os.environ.get("DATABASE_URL")
    if not database_url:
        raise RuntimeError(
            "DATABASE_URL is not set — saved-mapping storage is not configured. "
            "Set it to your Postgres connection string (see the Vercel Postgres/Neon "
            "integration's connection details)."
        )
    conn = psycopg.connect(database_url)
    global _schema_ready
    if not _schema_ready:
        _ensure_schema(conn)
        _schema_ready = True
    return conn


def _ensure_schema(conn):
    with conn.cursor() as cur:
        cur.execute(
            """
            CREATE TABLE IF NOT EXISTS mappings (
                name TEXT PRIMARY KEY,
                config JSONB NOT NULL,
                updated_at TIMESTAMPTZ NOT NULL DEFAULT now()
            )
            """
        )
    conn.commit()
    _seed_default_mapping(conn)


def _seed_default_mapping(conn):
    if not os.path.isfile(_DEFAULT_MAPPING_PATH):
        return
    with open(_DEFAULT_MAPPING_PATH, "r", encoding="utf-8") as f:
        config = json.load(f)
    with conn.cursor() as cur:
        cur.execute(
            "INSERT INTO mappings (name, config) VALUES (%s, %s) ON CONFLICT (name) DO NOTHING",
            (_DEFAULT_MAPPING_NAME, Jsonb(config)),
        )
    conn.commit()
    logger.info(
        "MAPPING STORE: ensured bundled default mapping '%s' exists (no-op if already present)",
        _DEFAULT_MAPPING_NAME,
    )


def list_mapping_names():
    with _connect() as conn:
        with conn.cursor() as cur:
            cur.execute("SELECT name FROM mappings ORDER BY name")
            return [row[0] for row in cur.fetchall()]


def get_mapping(name):
    """Returns the mapping's config dict, or None if no mapping exists under that name."""
    with _connect() as conn:
        with conn.cursor() as cur:
            cur.execute("SELECT config FROM mappings WHERE name = %s", (name,))
            row = cur.fetchone()
            return row[0] if row else None


def save_mapping(name, config):
    """Inserts or overwrites (upsert) the mapping under the given name."""
    with _connect() as conn:
        with conn.cursor() as cur:
            cur.execute(
                """
                INSERT INTO mappings (name, config, updated_at) VALUES (%s, %s, now())
                ON CONFLICT (name) DO UPDATE SET config = EXCLUDED.config, updated_at = now()
                """,
                (name, Jsonb(config)),
            )


def delete_mapping(name):
    """Deletes a mapping by name. Returns True if a row was deleted, False if it didn't exist."""
    with _connect() as conn:
        with conn.cursor() as cur:
            cur.execute("DELETE FROM mappings WHERE name = %s", (name,))
            return cur.rowcount > 0


def delete_all_mappings(except_name):
    """Deletes every saved mapping except the given (protected) name."""
    with _connect() as conn:
        with conn.cursor() as cur:
            cur.execute("DELETE FROM mappings WHERE name != %s", (except_name,))

File: backend-ts/src/mappingStore.ts
import { Collection, Document, MongoClient } from 'mongodb';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { createLogger } from './logger.js';

const logger = createLogger('mappingStore');

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// Bundled preset shipped in the repo -- seeded into the collection once so a
// fresh database starts with the same default mapping the app has always
// shipped.
const DEFAULT_MAPPING_NAME = 'electra_default';
const DEFAULT_MAPPING_PATH = path.join(__dirname, '..', 'mappings', `${DEFAULT_MAPPING_NAME}.json`);

interface MappingDocument extends Document {
  name: string;
  config: Record<string, unknown>;
  updatedAt: Date;
}

// A single MongoClient (with its own internal connection pool) is reused for
// the life of the process -- the idiomatic pattern for MongoDB in a
// long-running Node server, unlike the "new connection per call" pattern the
// Postgres version used specifically for serverless/pooled Postgres.
let cachedClient: MongoClient | null = null;
let indexEnsured = false;

async function getCollection(): Promise<Collection<MappingDocument>> {
  const uri = process.env.MONGODB_URI;
  if (!uri) {
    throw new Error(
      'MONGODB_URI is not set — saved-mapping storage is not configured. Set it to your MongoDB connection string.'
    );
  }
  if (!cachedClient) {
    cachedClient = new MongoClient(uri);
    await cachedClient.connect();
  }
  const collection = cachedClient.db().collection<MappingDocument>('mappings');
  if (!indexEnsured) {
    await collection.createIndex({ name: 1 }, { unique: true });
    await seedDefaultMapping(collection);
    indexEnsured = true;
  }
  return collection;
}

async function seedDefaultMapping(collection: Collection<MappingDocument>): Promise<void> {
  if (!fs.existsSync(DEFAULT_MAPPING_PATH)) return;
  const config = JSON.parse(fs.readFileSync(DEFAULT_MAPPING_PATH, 'utf-8'));
  await collection.updateOne(
    { name: DEFAULT_MAPPING_NAME },
    { $setOnInsert: { name: DEFAULT_MAPPING_NAME, config, updatedAt: new Date() } },
    { upsert: true }
  );
  logger.info(`MAPPING STORE: ensured bundled default mapping '${DEFAULT_MAPPING_NAME}' exists (no-op if already present)`);
}

export async function listMappingNames(): Promise<string[]> {
  const collection = await getCollection();
  const docs = await collection
    .find({}, { projection: { name: 1 } })
    .sort({ name: 1 })
    .toArray();
  return docs.map((d) => d.name);
}

/** Returns the mapping's config, or null if no mapping exists under that name. */
export async function getMapping(name: string): Promise<Record<string, unknown> | null> {
  const collection = await getCollection();
  const doc = await collection.findOne({ name });
  return doc ? doc.config : null;
}

/** Inserts or overwrites (upsert) the mapping under the given name. */
export async function saveMapping(name: string, config: Record<string, unknown>): Promise<void> {
  const collection = await getCollection();
  await collection.updateOne({ name }, { $set: { name, config, updatedAt: new Date() } }, { upsert: true });
}

/** Deletes a mapping by name. Returns true if a document was deleted, false if it didn't exist. */
export async function deleteMapping(name: string): Promise<boolean> {
  const collection = await getCollection();
  const result = await collection.deleteOne({ name });
  return result.deletedCount > 0;
}

/** Deletes every saved mapping except the given (protected) name. */
export async function deleteAllMappings(exceptName: string): Promise<void> {
  const collection = await getCollection();
  await collection.deleteMany({ name: { $ne: exceptName } });
}

/** Test-only escape hatch: forces a fresh connection on the next call, so
 * each test file can point at its own throwaway in-memory MongoDB instance. */
export async function _resetConnectionForTests(): Promise<void> {
  if (cachedClient) {
    await cachedClient.close();
    cachedClient = null;
  }
  indexEnsured = false;
}
